Docs: Glue Exporter and related (#6823)
treeverse/lakeFS: docs/integrations/glue_metastore.md

---
title: Glue / Athena support
description: This section explains how to query data from lakeFS branches in services backed by Glue Metastore in particular Athena.
parent: Integrations
redirect_from: /using/glue_metastore.html
---

# Using lakeFS with the Glue Catalog

{% include toc_2-3.html %}

## About Glue Metastore

[AWS Glue](https://docs.aws.amazon.com/glue/latest/dg/tables-described.html) has a metastore that can store metadata related to Hive and other services (such as Spark and Trino). It has metadata such as the location of the table, information about columns, partitions and much more.

## Support in lakeFS

The integration between Glue and lakeFS is based on [Data Catalog Exports]({% link howto/catalog_exports.md %}).

### What is supported 

> - Creating a uniqueue table in Glue Catalog per lakeFS repository / ref / commit. 

Reviewer: ```suggestion
- Creating a unique table in Glue Catalog per lakeFS repository / ref / commit. 
```

> ## Support in lakeFS
> 
> The integration between Glue and lakeFS is based on [Data Catalog Exports]({% link howto/catalog_exports.md %}).
> 
> ### What is supported 
> 
> - Creating a unique table in Glue Catalog per lakeFS repository / ref / commit. 
> - No data copying is required, the table location is a path to a symlinks structure in S3 based on Hive's [SymlinkTextInputFormat](https://svn.apache.org/repos/infra/websites/production/hive/content/javadocs/r2.1.1/api/org/apache/hadoop/hive/ql/io/SymlinkTextInputFormat.html) and the [table partitions](https://docs.aws.amazon.com/glue/latest/dg/tables-described.html#tables-partition) are maintained.
> - Tables are described via [Hive format in `_lakefs_tables/<my_table>.yaml`]({% link howto/catalog_exports.md %}#hive-tables).
> - Currently the data query in Glue metastore is Read-Only operation and mutating data requires writting to lakeFS and letting the export hook run.
> 
> ### How it works 
> 
> Based on event lakeFS events such as `post-commit` an Action will run a script that will create Symlink structures in S3 and then will register a table in Glue.
> The Table data location will be the generated Symlinks root path.
> 
> There are 4 key pieces:
> 
> 1. Table description at `_lakefs_tables/<your-table>.yaml`
> 2. Lua script that will do the export using [symlink_exporter]({% link howto/hooks/lua.md %}#lakefscatalogexportsymlink_exporter) and [glue_exporter]({% link howto/hooks/lua.md %}#lakefscatalogexportglue_exporter) packages.
> 3. [Action Lua Hook]({% link howto/catalog_exports.md %}#running-an-exporter) to execute the lua hook.
> 4. Write some lakeFS table data ([Spark]({% link integrations/spark.md %}), CSV, etc)
> 
> To learn more check [Data Catalog Exports]({% link howto/catalog_exports.md %}).

Reviewer 2: ```suggestion
This guide will show you how to use lakeFS with the Glue Data Catalog.
You'll be able to query your lakeFS data by specifying the branch, ref, or commit in your SQL query.
Currently, only read operations are supported on the tables.

You will set up the automation required to work with lakeFS on top of the Glue Data Catalog, including:
1. Create a table descriptor under `_lakefs_tables/<your-table>.yaml`. This will represent your table schema.
2. Write an exporter script that will:
   * Mirror your branch's state into symlink files readable by Athena
   * Export the table descriptors from your branch to the Glue Catalog.
3. Set up lakeFS hooks that will run the above script when specific events occur.
```

Author: Thanks! That's an awesome suggestion, did that with minor changes + added links.

## Example: Using Athena to query lakeFS data

> ### Pre-requisite 
> 
> 1. Glue Database to use (lakeFS does not create a database).
> 2. AWS Credentials with permission to manage Glue, Athena Query and S3 access.
> 3. lakeFS [Actions]({% link howto/hooks/index.md %}) enabled and configured with S3 blockstore.
> 
> ### Create lakeFS repository 
> 
> Let's create a repository named `catalogs`. 
> 
> ```bash
> lakectl repo create lakefs://catalogs s3://<lakefs-bucket>/<repo-prefix>
> ```

Reviewer 2: ```suggestion
### Prerequisites

Before starting, make sure you have:
1. An active lakeFS installation with S3 as the backing storage, and a repository in this installation.
2. A database in Glue Data Catalog (lakeFS does not create one).
3. AWS Credentials with permission to manage Glue, Athena Query and S3 access.
```

### Add table descriptor

Let's define a table and commit to lakeFS. 
Save the YAML below as `animals.yaml` and upload it to lakeFS. 

```bash
lakectl fs upload lakefs://catalogs/main/_lakefs_tables/animals.yaml -s ./animals.yaml && \
lakectl commit lakefs://catalogs/main -m "added table"
```

```yaml 
name: animals
type: hive
# data location root in lakeFS
path: tables/animals
# partitions order
partition_columns: ['type', 'weight']
schema:
  type: struct
  # all the columns spec
  fields:
    - name: type
      type: string
      nullable: true
      metadata:
        comment: axolotl, cat, dog, fish etc
    - name: weight
      type: integer
      nullable: false
      metadata: {}
    - name: name
      type: string
      nullable: false
      metadata: {}
```

### Write some table data

> Add some table data. 

Reviewer 2: I suggest to keep this section short, even a one-liner if that's possible. It's not in the scope of the article, it's more like a means to an end.
It can be something like "Insert data under the table path, using your preferred method. For example, here is how to write data using Spark: ..."

The commit bit doesn't need to be in the code, you can simply say "then perform a commit in lakeFS".

Author: @itaiad200 @arielshaqed opinions?

Reviewer: @johnnyaug is right that this doesn't belong in "integrations".  Unfortunately if you remove this text, it means you've just been volunteered to write the blog post / tech note.

Author: @arielshaqed so setting aside volunteering, you agree that this step by step tutorial on how to write the data itself should be removed to a general guidance? :)

Author: @johnnyaug I took your advice and removed 99% of the content from there!

configuration.verify_ssl = False
lakefsApi = LakeFSClient(configuration)
# init spark session 
builder = SparkSession.builder.appName("MyApp") \
    .config("spark.hadoop.fs.s3a.access.key", lakefs_access_key_id) \
    .config("spark.hadoop.fs.s3a.secret.key",lakefs_secret_access_key) \
    .config("spark.hadoop.fs.s3a.path.style.access", "true") \
    .config("spark.hadoop.fs.s3a.endpoint", lakefs_endpoint) \
    .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
spark = builder.getOrCreate()
# write table data as parquet files to lakeFS 
repo="catalogs"
ref = "main"
tablePath = f"s3a://{repo}/{ref}/tables/animals"
data=[
    ("axolotl", 10, "jayson"),
    ("axolotl", 10, "john"),
    ("axolotl", 12, "mark"),
    ("turtle", 12, "mrgreen"),
    ("bear", 30, "winnie"),
]
columns=["type","weight","name"]
df=spark.createDataFrame(data,columns)
df.write.partitionBy("type","weight").mode("overwrite").parquet(f"{tablePath}/data.parquet")
# commit data to lakeFS 
lakefsApi.commits.commit(repository=repo,branch=ref,commit_creation=CommitCreation(message="data"))
```
  </div>

</div>

> ### Add Glue Exporter
> 
> The current step requires 2 things: 
> 1. Lua script.
> 2. Action to trigger the script.
> 
> #### Lua packages (Exporters)
> 
> - [symlink_exporter]({% link howto/hooks/lua.md %}#lakefscatalogexportsymlink_exporter)
> - [glue_exporter]({% link howto/hooks/lua.md %}#lakefscatalogexportglue_exporter)
> 
> #### Exporter Script 
> 
> ##### 1. Create Lua script:
>   
> For the simple strategy of creating a glue table per repo / branch / commit we can simply copy-paste the following script and re-use it.
> Upload the script to `scripts/animals_exporter.lua` (could be any path).

Reviewer 2: ```suggestion
### The exporter script

Upload the following script to your main branch under `scripts/animals_exporter.lua` (or a path of your choice).
```

Author: like it! but with a little twist I kept the libraries reference.

```lua 
local aws = require("aws")
local symlink_exporter = require("lakefs/catalogexport/symlink_exporter")
local glue_exporter = require("lakefs/catalogexport/glue_exporter")
-- settings 
local access_key = args.aws.aws_access_key_id
local secret_key = args.aws.aws_secret_access_key
local region = args.aws.aws_region
local table_path = args.table_source -- table descriptor 
local db = args.catalog.db_name -- glue db
local table_input = args.catalog.table_input -- table input (AWS input spec) for Glue
-- export symlinks 
local s3 = aws.s3_client(access_key, secret_key, region)
local result = symlink_exporter.export_s3(s3, table_path, action, {debug=true})
-- register glue table
local glue = aws.glue_client(access_key, secret_key, region)
local res = glue_exporter.export_glue(glue, db, table_path, table_input, action, {debug=true})
```

> ##### 2. Configure Action Hooks:

Reviewer 2: ```suggestion
### Configure Action Hooks
```

        table_source: '_lakefs_tables/animals.yaml'
      script: |
        local exporter = require("lakefs/catalogexport/symlink_exporter")
        local aws = require("aws")
        local table_path = args.table_source
        local s3 = aws.s3_client(args.aws.aws_access_key_id, args.aws.aws_secret_access_key, args.aws.aws_region)
        exporter.export_s3(s3, table_path, action, {debug=true})
  - id: glue_exporter
    type: lua
    properties:
      args:
        aws:
          aws_access_key_id: "<AWS_ACCESS_KEY_ID>"
          aws_secret_access_key: "<AWS_SECRET_ACCESS_KEY>"
          aws_region: "<AWS_REGION>"
        table_source: '_lakefs_tables/animals.yaml'
        catalog:
          db_name: "my-glue-db"
          table_input: # add glue table input here 
      script: |
        local aws = require("aws")
        local exporter = require("lakefs/catalogexport/glue_exporter")
        local glue = aws.glue_client(args.aws.aws_access_key_id, args.aws.aws_secret_access_key, args.aws.aws_region)
        exporter.export_glue(glue, args.catalog.db_name, args.table_source, args.catalog.table_input, action, {debug=true})  
```

  </div>
</div>

> Once we are adding the script and the action file we can commit it and since the action is configured to `post-commit` it will run after the commit! 

Reviewer: ```suggestion
Adding the script and the action file to the repository and commit it.  This is a post-commit action - it will run after the commit! 
```

```bash
lakectl fs upload lakefs://catalogs/main/scripts/animals_exporter.lua -s ./animals_exporter.lua
lakectl fs upload lakefs://catalogs/main/_lakefs_actions/animals_glue.yaml -s ./animals_glue.yaml
lakectl commit lakefs://catalogs/main -m "trigger first export hook"
```

> Once the action finished running we can see the logs the result.

Reviewer: ```suggestion
After the action finishes running we can see the result in the action logs.
```
